File: src-tauri/sidecars/arrow-maxcompute/ArrowSidecar.java
import com.aliyun.odps.Odps;
import com.aliyun.odps.account.AliyunAccount;
import com.aliyun.odps.tunnel.TableTunnel;
import com.aliyun.odps.tunnel.TableTunnel.DownloadSession;
import com.aliyun.odps.data.ArrowRecordReader;
import com.aliyun.odps.thirdparty.org.apache.arrow.memory.BufferAllocator;
import com.aliyun.odps.thirdparty.org.apache.arrow.memory.RootAllocator;
import com.aliyun.odps.thirdparty.org.apache.arrow.vector.VectorSchemaRoot;
import com.aliyun.odps.thirdparty.org.apache.arrow.vector.ipc.ArrowStreamWriter;

import java.io.BufferedOutputStream;
import java.io.OutputStream;

/**
 * MaxCompute Arrow sidecar: bulk-download a table via the ODPS SDK's
 * `TableTunnel` + `ArrowTunnelRecordReader` (no instance-tunnel 10000-row cap),
 * re-serializing each batch as an Arrow IPC STREAM to stdout (consumed by the
 * Rust host's `arrow::ipc::reader::StreamReader` → DuckDB `appender-arrow`).
 * All logs go to stderr.
 *
 *   usage: ArrowSidecar <project.table> [start] [count]
 *   env  : ODPS_ENDPOINT, ODPS_ACCESS_KEY_ID, ODPS_ACCESS_KEY_SECRET,
 *          [ODPS_TUNNEL_ENDPOINT], [ODPS_REGION], [ODPS_PROJECT]
 *
 * `count == 0` (or beyond EOF) means "to end of table". The Rust host launches
 * this with `--add-opens=java.base/java.nio=ALL-UNNAMED ...` and
 * `-XX:MaxDirectMemorySize=8G` (see `external/arrow_sidecar.rs` ADD_OPENS).
 *
 * Memory note: a single session can accumulate ~250 bytes/row of Arrow direct
 * memory; `-XX:MaxDirectMemorySize=8G` covers ~30M rows. Larger tables need
 * windowed pulls (future: per-window allocator + `getRawStream()` bypass) —
 * see `spike/REPORT.md`.
 */
public class ArrowSidecar {
    public static void main(String[] args) throws Exception {
        String endpoint = System.getenv("ODPS_ENDPOINT");
        String akId = System.getenv("ODPS_ACCESS_KEY_ID");
        String akSecret = System.getenv("ODPS_ACCESS_KEY_SECRET");
        String tunnel = System.getenv("ODPS_TUNNEL_ENDPOINT");
        String region = System.getenv("ODPS_REGION");
        if (endpoint == null || akId == null || akSecret == null || args.length < 1) {
            System.err.println("usage: ArrowSidecar <project.table> [start] [count]  "
                + "(env: ODPS_ENDPOINT, ODPS_ACCESS_KEY_ID/SECRET, [ODPS_TUNNEL_ENDPOINT], [ODPS_REGION])");
            System.exit(2);
        }
        String full = args[0];
        String project, table;
        int dot = full.indexOf('.');
        if (dot < 0) { project = System.getenv("ODPS_PROJECT"); table = full; }
        else { project = full.substring(0, dot); table = full.substring(dot + 1); }

        AliyunAccount acct = new AliyunAccount(akId, akSecret);
        if (region != null && !region.isEmpty()) acct.setRegion(region);
        Odps odps = new Odps(acct);
        odps.setEndpoint(endpoint.replaceFirst("/+$", ""));
        odps.setDefaultProject(project);
        if (tunnel != null && !tunnel.isEmpty()) odps.setTunnelEndpoint(tunnel);
        System.err.println("[arrow] project=" + project + " table=" + table);

        TableTunnel t = new TableTunnel(odps);
        long tCreate = System.currentTimeMillis();
        DownloadSession session = t.createDownloadSession(project, table);
        long total = session.getRecordCount();
        long start = 0;
        long count = total;
        if (args.length >= 3) {            // <table> <start> <count>
            start = Long.parseLong(args[1]);
            count = Long.parseLong(args[2]);
        } else if (args.length >= 2) {     // <table> <count>  (from 0, backward-compat)
            count = Long.parseLong(args[1]);
        }
        if (start < 0) start = 0;
        if (count <= 0 || start + count > total) count = total - start;
        System.err.println("[arrow] createDownloadSession ok in "
            + (System.currentTimeMillis() - tCreate) + "ms; recordCount=" + total
            + " start=" + start + " pulling=" + count);

        BufferAllocator alloc = new RootAllocator(Long.MAX_VALUE);
        ArrowRecordReader reader = session.openArrowRecordReader(start, count, alloc);

        OutputStream out = new BufferedOutputStream(System.out, 1 << 20);
        long rows = 0, batches = 0;
        long t0 = System.currentTimeMillis();
        VectorSchemaRoot root = reader.read();
        if (root == null) {
            System.err.println("[arrow] no data; exiting");
            out.flush();
            reader.close();
            return;
        }
        ArrowStreamWriter writer = new ArrowStreamWriter(root, null, out);
        writer.start();
        do {
            rows += root.getRowCount();
            batches++;
            writer.writeBatch();
            if (batches % 50 == 0) {
                System.err.println("[arrow] batches=" + batches + " rows=" + rows
                    + " (" + (rows * 1000L / Math.max(1, System.currentTimeMillis() - t0)) + " rows/s)");
            }
        } while ((root = reader.read()) != null);
        writer.end();
        writer.close();
        out.flush();
        reader.close();
        long elapsed = System.currentTimeMillis() - t0;
        double rps = elapsed > 0 ? rows * 1000.0 / elapsed : 0;
        System.err.println("[arrow] DONE batches=" + batches + " rows=" + rows
            + " elapsed=" + elapsed + "ms  ->  " + String.format("%.0f", rps) + " rows/sec");
    }
}

File: src-tauri/src/external/arrow_sidecar.rs
//! MaxCompute bulk download via the Arrow sidecar: ODPS `TableTunnel` +
//! `ArrowTunnelRecordReader` → Arrow IPC stream → DuckDB `appender-arrow`.
//!
//! This is the MaxCompute-specific "fast lane": no instance-tunnel 10000-row
//! cap, no RecordPack binary decode, columnar zero-copy ingest. Validated in
//! the spike (~17k rows/sec single-session, ~64k at 10-way concurrency; see
//! `spike/REPORT.md`).

use std::io::BufReader;
use std::process::{Command, Stdio};
use std::time::Instant;

use arrow::datatypes::DataType;
use arrow::ipc::reader::StreamReader;
use serde::Serialize;

use crate::db::{DbConnectionRecord, MaxcomputeOpts};

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PullStats {
    pub rows: u64,
    pub batches: u64,
    pub elapsed_secs: f64,
    pub rows_per_sec: f64,
}

/// JVM flags required by the shaded Arrow `MemoryUtil` (reflective access to
/// `java.nio.Buffer.address`) on JDK 17+, plus a generous direct-memory ceiling.
/// Windowed pulls + per-window `allocator.close()` keep direct memory bounded.
const ADD_OPENS: &[&str] = &[
    "--add-opens=java.base/java.io=ALL-UNNAMED",
    "--add-opens=java.base/java.lang=ALL-UNNAMED",
    "--add-opens=java.base/java.lang.reflect=ALL-UNNAMED",
    "--add-opens=java.base/java.net=ALL-UNNAMED",
    "--add-opens=java.base/java.nio=ALL-UNNAMED",
    "--add-opens=java.base/java.nio.charset=ALL-UNNAMED",
    "--add-opens=java.base/java.util=ALL-UNNAMED",
    "--add-opens=java.base/java.util.concurrent=ALL-UNNAMED",
    "--add-opens=java.base/jdk.internal.misc=ALL-UNNAMED",
    "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED",
    "-XX:MaxDirectMemorySize=8G",
];

/// Map an Arrow `DataType` to a DuckDB column type for `CREATE TABLE`.
/// Complex types (List/Struct/Map/FixedSizeList) fall back to VARCHAR for now
/// (the ODPS SDK's Arrow accessors cover them; full mapping is a follow-up).
fn arrow_to_duckdb_type(t: &DataType) -> &'static str {
    match t {
        DataType::Boolean => "BOOLEAN",
        DataType::Int8 | DataType::Int16 | DataType::Int32 => "INTEGER",
        DataType::Int64 | DataType::UInt64 => "BIGINT",
        DataType::UInt8 | DataType::UInt16 | DataType::UInt32 => "INTEGER",
        DataType::Float16 | DataType::Float32 => "FLOAT",
        DataType::Float64 => "DOUBLE",
        DataType::Utf8 | DataType::LargeUtf8 => "VARCHAR",
        DataType::Binary | DataType::LargeBinary => "BLOB",
        DataType::Date32 | DataType::Date64 => "DATE",
        DataType::Timestamp(_, _) => "TIMESTAMP",
        DataType::Time32(_) | DataType::Time64(_) => "TIME",
        DataType::Decimal128(_, _) | DataType::Decimal256(_, _) => "DECIMAL",
        DataType::Duration(_) => "BIGINT",
        _ => "VARCHAR",
    }
}

/// Pull a MaxCompute table (or a `[start, start+count)` row window of it, or a
/// single partition of a partitioned table) through the Arrow sidecar and
/// ingest it into DuckDB as `local_table`.
///
/// - `table_ref`: fully-qualified `project.table` (the sidecar resolves the
///   session's default project from `ODPS_PROJECT`; a bare name also works).
/// - `sidecar_jar`: path to `arrow-maxcompute-sidecar.jar` (a bundled resource).
/// - `driver_jars`: resolved vendor JARs (odps-sdk-core etc.) — runtime classpath.
/// - `start` / `count`: row window; `count == 0` means "to end of table/partition".
/// - `partition_spec`: when `Some`, download only that partition (e.g.
///   `ds=20250701` or multi-level `ds=20250701/region=cn`). Required for
///   partitioned tables; ignored for non-partitioned tables.
///
/// AK/SK are passed via the child env (`ODPS_ACCESS_KEY_ID/SECRET`), never logged.
pub fn pull_table(
    duck_conn: &duckdb::Connection,
    rec: &DbConnectionRecord,
    opts: &MaxcomputeOpts,
    table_ref: &str,
    local_table: &str,
    sidecar_jar: &str,
    driver_jars: &[String],
    start: u64,
    count: u64,
    partition_spec: Option<&str>,
) -> Result<PullStats, String> {
    let java = crate::external::jdbc_sidecar::find_java_bin()
        .ok_or_else(|| "未找到 Java 运行时（MaxCompute 物化需要 JRE 17+）".to_string())?;

    // classpath = driver jars : sidecar jar
    let mut cp = driver_jars.join(":");
    if !cp.is_empty() {
        cp.push(':');
    }
    cp.push_str(sidecar_jar);

    let t0 = Instant::now();
    let mut cmd = Command::new(&java);
    cmd.args(ADD_OPENS)
        .arg("-cp")
        .arg(&cp)
        .arg("ArrowSidecar")
        .arg(table_ref)
        .arg(start.to_string())
        .arg(count.to_string());
    // Optional 4th arg: partition spec for partitioned tables.
    if let Some(ps) = partition_spec {
        if !ps.is_empty() {
            cmd.arg(ps);
        }
    }
    let mut child = cmd
        .env("ODPS_ENDPOINT", &opts.endpoint)
        .env("ODPS_ACCESS_KEY_ID", rec.username.as_str())
        .env("ODPS_ACCESS_KEY_SECRET", rec.password.as_str())
        .env("ODPS_PROJECT", &opts.project)
        .stdout(Stdio::piped())
        .stderr(Stdio::inherit())
        .spawn()
        .map_err(|e| format!("启动 Arrow sidecar 失败: {e}"))?;

    let stdout = child.stdout.take().ok_or("arrow sidecar 无 stdout")?;
    let reader = BufReader::with_capacity(1 << 20, stdout);
    let mut stream = StreamReader::try_new(reader, None)
        .map_err(|e| format!("打开 Arrow IPC 流失败: {e}"))?;
    let schema = stream.schema().clone();

    // CREATE TABLE IF NOT EXISTS from the Arrow schema. For a fresh pull this
    // creates the table; for subsequent partition/window pulls into the same
    // local table, the IF NOT EXISTS makes it a no-op (the appender below
    // appends to the existing table). This lets the caller drive multiple
    // pull_table() calls into one local table without managing DDL itself.
    let cols: Vec<String> = schema
        .fields()
        .iter()
        .map(|f| format!("\"{}\" {}", f.name().replace('"', "\"\""), arrow_to_duckdb_type(f.data_type())))
        .collect();
    let create_sql = format!("CREATE TABLE IF NOT EXISTS \"{local_table}\" ({});", cols.join(", "));
    duck_conn
        .execute(&create_sql, [])
        .map_err(|e| format!("建本地表失败: {e}"))?;

    let mut app = duck_conn
        .appender(local_table)
        .map_err(|e| format!("打开 appender 失败: {e}"))?;
    let mut rows: u64 = 0;
    let mut batches: u64 = 0;
    let mut last_log = Instant::now();
    while let Some(batch) = stream.next() {
        let batch = batch.map_err(|e| format!("读取 Arrow batch 失败: {e}"))?;
        rows += batch.num_rows() as u64;
        batches += 1;
        app.append_record_batch(batch)
            .map_err(|e| format!("写入 DuckDB 失败: {e}"))?;
        // Log progress every 5s so the operator can see the pull is alive.
        if last_log.elapsed().as_secs() >= 5 {
            let elapsed = t0.elapsed().as_secs_f64();
            let rps = if elapsed > 0.0 { rows as f64 / elapsed } else { 0.0 };
            tracing::info!(
                category = "link",
                "maxcompute: pulling {table_ref} -> {local_table}: {rows} rows, {batches} batches, {elapsed:.1}s ({rps:.0} rows/s)"
            );
            last_log = Instant::now();
        }
    }
    drop(app);
    let _ = child.wait();
    let elapsed = t0.elapsed().as_secs_f64();
    let rps = if elapsed > 0.0 { rows as f64 / elapsed } else { 0.0 };
    Ok(PullStats { rows, batches, elapsed_secs: elapsed, rows_per_sec: rps })
}
